rclcpp/test/rclcpp/test_client.cpp

```diff
@@ -340,3 +341,105 @@ TEST_F(TestClientWithServer, take_response) {
       rclcpp::exceptions::RCLError);
   }
 }
+
+TEST_F(TestClient, rcl_client_request_publisher_get_actual_qos_error) {
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_client_request_publisher_get_actual_qos, nullptr);
+  auto client = node->create_client<test_msgs::srv::Empty>("service");
+  RCLCPP_EXPECT_THROW_EQ(
+    client->get_request_publisher_actual_qos(),
+    std::runtime_error("failed to get client's request publisher qos settings: error not set"));
+}
+
+TEST_F(TestClient, rcl_client_response_subscription_get_actual_qos_error) {
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_client_response_subscription_get_actual_qos, nullptr);
+  auto client = node->create_client<test_msgs::srv::Empty>("service");
+  RCLCPP_EXPECT_THROW_EQ(
+    client->get_response_subscription_actual_qos(),
+    std::runtime_error("failed to get client's response subscription qos settings: error not set"));
+}
+
+TEST_F(TestClient, client_qos) {
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_services_default;
+  qos_profile.liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+  uint64_t duration = 1;
+  qos_profile.deadline = {duration, duration};
+  qos_profile.lifespan = {duration, duration};
+  qos_profile.liveliness_lease_duration = {duration, duration};
+
+  auto client =
+    node->create_client<test_msgs::srv::Empty>("client", qos_profile);
+
+  auto init_qos =
+    rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(qos_profile), qos_profile);
+  auto rp_qos = client->get_request_publisher_actual_qos();
+  auto rs_qos = client->get_response_subscription_actual_qos();
+
+  EXPECT_EQ(init_qos, rp_qos);
+  // Lifespan has no meaning for subscription/readers
+  rs_qos.lifespan(qos_profile.lifespan);
+  EXPECT_EQ(init_qos, rs_qos);
+}
+
+TEST_F(TestClient, client_qos_depth) {
```

Comment:
This test looks like too complex for what rclcpp does (?).
It depends on the rmw implementation behaving correctly.

Reply:
It's not bad to keep it though, but currently this kind of tests live in `test_rclcpp` or `test_communication`.